### PR TITLE
Add a Resources Page to the Community Section

### LINKED ISSUE #2041: Update Data Model to Allow Recording Resources from the Community
Thank you for the glorious work on ocaml.org!

The learning resources are great already but may I suggest adding as extra material the online exercises available there:

https://ocaml-sf.org/learn-ocaml-public/#activity=exercises

? 

These exercises are probably interesting for beginners and intermediate OCaml developers.

### PR DESCRIPTION
**Pull request by 0xrotense**
Fixes: #2041

> Sure!
> 
> One possible way to go is to define another category of served data. This would allow defining a tile-base UI (like this one https://ocaml.org/jobs). We'd have one tile per learn resource. Learn-OCaml would be one. There's also this:
> 
> https://github.com/gs0510/ofronds
> https://github.com/sudha247/learn-ocaml-workshop
> https://github.com/kayceesrk/cs3100_f19
> I suspect there's even more.
> 
> This implies creating a new kind of served data. Folowing-up with the jobs example, this file: [data/jobs](https://github.com/ocaml/ocaml.org/blob/main/data/jobs.yml) entails the jobs page.
> 
> In [tool/ood-gen/lib/job.ml](https://github.com/ocaml/ocaml.org/blob/main/tool/ood-gen/lib/job.ml) there's the Yaml parser. It translates into OCaml sources which are compiled and linked (ocaml.org runs from memory).
> In [src/ocamlorg_frontend/pages/jobs.eml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_frontend/pages/jobs.eml) there's the HTML template which is turned into the paged served with the Yaml data
> In [src/ocamlorg_web/lib/handler.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_web/lib/handler.ml) and [src/ocamlorg_web/lib/router.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_web/lib/router.ml) there's the bits that glues everything together. There's also some stuff in [src/ocamlorg_data/lib/data.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_data/data.ml)
> That's the overall logic. @Clairevanden will think of the design.
> 
> Regarding the Yaml data we probably need fields for:
> 
> Title
> Short description
> Long description
> Image
> Online Url
> Source Url
> License
> This whole thing would allow displaying the tiles either in the exercises, in the community or anywhere else.

Proposed Designs by @sabine: https://www.figma.com/file/7hmoWkQP9PgLTfZCqiZMWa/OCaml-Community-Pages?type=design&node-id=152%3A386&mode=dev&t=kePoh4Cvcc35Nemo-1

To Do:

- [x] Create a new kind of served data, e.g., `data/resources.yml`, that will entail the new resources page. These are the fields needed for the YAML DATA: [ddfb90f](https://github.com/ocaml/ocaml.org/pull/2184/commits/ddfb90f9a9bf3756a3363c742ec53450bd6282db)
       - Title
       - Short description
       - Long description
       - Image
       - Online Url
       - Source Url
       - License
- [x] Create and add the YAML parser to `tool/ood-gen/lib/resource.ml`.  It translates into OCaml sources which are compiled and linked (ocaml.org runs from memory). [78ab3bd](https://github.com/ocaml/ocaml.org/pull/2184/commits/78ab3bd6b31527c8e0666cb3ca99aa6a7383f6d8)
- [x] Create and add the HTML Template to `src/ocamlorg_frontend/pages/resources.eml`. There's the HTML template which is turned into the paged served with the Yaml data.
- [x] Lastly, Glue everything together in the following:
       - `src/ocamlorg_web/lib/handler.ml`
       - `src/ocamlorg_web/lib/router.ml`
       - `src/ocamlorg_data/lib/data.ml`

### CODE REVIEW

**Comment by 0xrotense** [comment]
Hello, @cuihtlauac, @shakthimaan and @SaySayo ,

Please help me review this to see if I'm on the right track.

Regards.

File: data/resources.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    source_url: https://github.com/sudha247/learn-ocaml-workshop
+    license: Apache License
+  - title: |-
+        CS3100: Paradigms of Programming (IITM Fall 2019)
+    short_description: This course is about the different paradigms of programming, their underlying concepts, and the relationships between them.
+    long_description: A course offered at the Computer Science and Engineering department of IIT Madras (CSE IITM) during the Fall semester of 2019. This contains course materials such as assignments, lecture notes, and other educational resources related to the course and may also be beneficial for others interested in learning about different programming paradigms.
+    image: https://ocaml.org/_/MTE3YWE5YzJlZmExNmYyMmI3ZmEwYjliMDEwNDBkMDY/logo-with-name-white.svg
+    online_url: https://kcsrk.info/cs3100_f19/
+    source_url: https://github.com/kayceesrk/cs3100_f19
+    license: MIT License
```
Comment (cuihtlauac) [inline data/resources.yml]: Add newline at end of file

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

**Review by christinerose — APPROVED** [review]

File: data/resources.yml (hunk 1 above)
Comment (0xrotense) [reply data/resources.yml]: Thanks for taking note of that, Now added.

**Comment by 0xrotense** [comment]
Please help me review the second task I just pushed.

@cuihtlauac, @shakthimaan and @SaySayo

Regards.

**Comment by 0xrotense** [comment]
Hello @cuihtlauac @SaySayo ,

I got this error when I added the `resource` handler to [src/ocamlorg_web/lib/handler.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_web/lib/handler.ml) and running `dune build`

Screenshot 2024-03-14 182426

Please what does the error mean and how do i go about it?  or is it because I have not updated the [src/ocamlorg_data/lib/data.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_data/data.ml) ?

Regards.

**Review by christinerose — APPROVED** [review]

**Comment by 0xrotense** [comment]
> Hello @cuihtlauac @SaySayo ,
> 
> I got this error when I added the `resource` handler to [src/ocamlorg_web/lib/handler.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_web/lib/handler.ml) and running `dune build`
> 
> Screenshot 2024-03-14 182426
> Please what does the error mean and how do i go about it? or is it because I have not updated the [src/ocamlorg_data/lib/data.ml](https://github.com/ocaml/ocaml.org/blob/main/src/ocamlorg_data/data.ml) ?
> 
> Regards.

kindly help review this @cuihtlauac

**Comment by cuihtlauac** [comment]
@0xrotense Sorry I missed your message yesterday.

That's because you haven't added the type resource in `src/ocamlorg_data/data.ml` and `src/ocamlorg_date/data.mli`. Look into those files how Jobs are handled. You need something similar for resources.

**Comment by 0xrotense** [comment]
> @0xrotense Sorry I missed your message yesterday.
> 
> That's because you haven't added the type resource in `src/ocamlorg_data/data.ml` and `src/ocamlorg_date/data.mli`. Look into those files how Jobs are handled. You need something similar for resources.

okay, thanks will look into that.

**Comment by 0xrotense** [comment]
> image
This is resources handler in `handler.ml` .

> Screenshot 2024-03-15 162425
The type resource in `data.ml`

> image
The Error I got just now after running `dune build`.

> image
The module interface in `data.mli`

@cuihtlauac

**Comment by 0xrotense** [comment]
> > image
> 
> This is resources handler in `handler.ml` .
> 
> > Screenshot 2024-03-15 162425
> 
> The type resource in `data.ml`
> 
> > image
> 
> The Error I got just now after running `dune build`.
> 
> > image
> 
> The module interface in `data.mli`
> 
> @cuihtlauac

Hi @cuihtlauac , Please help look into this.

Thanks

File: data/resources.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+resources:
+  - title:  Learn-OCaml
+    short_description: This is Learn-OCaml, a platform for learning the OCaml language, featuring a Web toplevel, an exercise environment.
+    long_description: A platform designed for learning the OCaml programming language. It features a web-based interactive toplevel, an environment for exercises, and a collection of lessons and tutorials. The platform is suitable for both beginners and experienced programmers and can be used for academic purposes or self-study.
```
Comment (erikmd) [inline data/resources.yml]: Just a small suggestion, to highlight one of the most typical learn-ocaml feature:

```suggestion
    long_description: A platform designed for learning the OCaml programming language. It features a web-based interactive toplevel, an environment for exercises with automated grading tests, and a collection of lessons and tutorials. The platform is suitable for both beginners and experienced programmers and can be used for academic purposes or self-study.
```

File: data/resources.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    license: Apache License
+  - title: |-
+        CS3100: Paradigms of Programming (IITM Fall 2019)
+    short_description: This course is about the different paradigms of programming, their underlying concepts, and the relationships between them.
+    long_description: A course offered at the Computer Science and Engineering department of IIT Madras (CSE IITM) during the Fall semester of 2019. This contains course materials such as assignments, lecture notes, and other educational resources related to the course and may also be beneficial for others interested in learning about different programming paradigms.
+    image: https://ocaml.org/_/MTE3YWE5YzJlZmExNmYyMmI3ZmEwYjliMDEwNDBkMDY/logo-with-name-white.svg
+    online_url: https://kcsrk.info/cs3100_f19/
+    source_url: https://github.com/kayceesrk/cs3100_f19
+    license: MIT License
+
```
Comment (cuihtlauac) [inline data/resources.yml]:
```suggestion
    license: MIT License

```

**Review by cuihtlauac — DISMISSED** [review]

**Comment by cuihtlauac** [comment]
@0xrotense 

You must edit the files `src/ocamlorg_data/data.ml` and `src/ocamlorg_data/data.mli`.

Essentially, the type defined in the `tool/ood-gen/lib/ressource.ml` must be copied in both places. With the wrapping into a submodule.

**Review by christinerose — APPROVED** [review]

**Review by christinerose — APPROVED** [review]

**Review by sabine — APPROVED** [review]
Hey @0xrotense, thanks for starting this work with this contribution! :heart:

I took a moment and finished the rest of the tasks, so we can merge this now! :sparkles:

**Pull request merged**